### PR TITLE
Move CapacityBuffersFakePodsRegistry to AutoscalingContext

### PR DESCRIPTION
**Pull request by walidghallab**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Move CapacityBuffersFakePodsRegistry to AutoscalingContext

Reasons:
- Allow all parties in CA loop to be able to read it.
- When multiple capacitybuffer strategies are used in different
  processors, this allow them to use the same registry and unify the
clearing of this registry to be in a single place in CA loop (instead of
being done in pod list processors).
- Avoid passing it around in code.

Since it is being part of AutoscalingContext, additional refactoring was done in order for it to have a cleaner and more extensible interface.

#### Which issue(s) this PR fixes:
Fixes #

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
-->
```docs

```

### CODE REVIEW

File: cluster-autoscaler/context/autoscaling_context.go [hunk 1]
```diff
@@ -145,19 +148,20 @@ func NewAutoscalingContext(
...
+		ClusterSnapshot:                 clusterSnapshot,
+		ExpanderStrategy:                expanderStrategy,
+		ProcessorCallbacks:              processorCallbacks,
+		DebuggingSnapshotter:            debuggingSnapshotter,
+		RemainingPdbTracker:             remainingPdbTracker,
+		ClusterStateRegistry:            clusterStateRegistry,
+		DraProvider:                     draProvider,
+		TemplateNodeInfoRegistry:        templateNodeInfoRegistry,
+		CsiProvider:                     csiProvider,
+		CapacityBuffersFakePodsRegistry: registry.NewFakePodsRegistry(nil),
```
Comment (norbertcyran) [inline cluster-autoscaler/context/autoscaling_context.go]: opinion: I think we are heavily overusing autoscaling context. We treat it as a god object storing all possible dependencies. That makes it unclear what are the real dependencies of a struct. Before this change, it was very explicit that `CapacityBufferPodListProcessor` depends on fake pods registry, now it's hidden in the implementation. It also adds some overhead in the test setup - previously we could just pass nil to `autoscalingCtx`, now we have to initialize it.

Personally, I'd very much prefer keeping the dependencies explicit. `AutoscalingContext` sounds to me like a short-lived data scoped to the specific autoscaler loop, not a container for dependency injection. @towca @jackfrancis WDYT?

File: cluster-autoscaler/capacitybuffer/fakepods/registry.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
+/*
```
Comment (norbertcyran) [inline cluster-autoscaler/capacitybuffer/fakepods/registry.go:1]: nit: how about moving this to `capacitybuffer/fakepods` pkg as `registry.go`?

File: cluster-autoscaler/capacitybuffer/fakepods/registry.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+package registry
+
+import (
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/autoscaler/cluster-autoscaler/apis/capacitybuffer/autoscaling.x-k8s.io/v1beta1"
+)
+
+// FakePodsRegistry tracks the relationship between fake pods (created for capacity buffers)
+// and their originating CapacityBuffer objects.
+// IMPORTANT: It is currently used only in the main CA loop and is not thread-safe.
```
Comment (norbertcyran) [inline cluster-autoscaler/capacitybuffer/fakepods/registry.go]: Please make it thread safe - it's a low effort change, and potentially we could save the pods to the registry in the controller (which runs in a separate goroutine) after resolving the pod template

File: cluster-autoscaler/capacitybuffer/registry/fake_pods_registry.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	return r.fakePodsUIDToBuffer[fakePodUID]
+}
+
+// SetCapacityBuffer registers a mapping between a fake pod's UID and the CapacityBuffer it was created from.
+func (r *FakePodsRegistry) SetCapacityBuffer(fakePodUID types.UID, buffer *v1beta1.CapacityBuffer) {
+	r.fakePodsUIDToBuffer[fakePodUID] = buffer
+}
+
+// ClearCapacityBuffer removes the mapping for the specified fake pod UID.
+func (r *FakePodsRegistry) ClearCapacityBuffer(fakePodUID types.UID) {
```
Comment (norbertcyran) [inline cluster-autoscaler/capacitybuffer/registry/fake_pods_registry.go]: nit: `Delete` or `Unset` would better match the underlying map semantics, and it would make it clearer that it's different from `Clear`

File: cluster-autoscaler/processors/capacitybuffer/pod_list_processor.go [hunk 5]
```diff
@@ -125,20 +107,13 @@ func (p *CapacityBufferPodListProcessor) Process(autoscalingCtx *ca_context.Auto
...
 	}
 	for _, fakePod := range fakePods {
-		p.buffersRegistry.FakePodsUIDToBuffer[string(fakePod.UID)] = buffer
-	}
-}
-
-func (p *CapacityBufferPodListProcessor) clearCapacityBufferRegistry() {
-	if p.buffersRegistry == nil {
-		return
+		bufferRegistry.SetCapacityBuffer(fakePod.UID, buffer)
```
Comment (norbertcyran) [inline cluster-autoscaler/processors/capacitybuffer/pod_list_processor.go:115]: general note: do we even need that registry? Couldn't we just generate fake pods with an owner reference to the corresponding buffer?

File: cluster-autoscaler/context/autoscaling_context.go (hunk 1 above)
Comment (walidghallab) [reply cluster-autoscaler/context/autoscaling_context.go]: Honestly while I agree about the general statements, I have different opinion about the interpretations. 

Regarding "AutoscalingContext sounds to me like a short-lived data scoped to the specific autoscaler loop", this is true, and that's exactly why I moved the registry there. Registry is short-lived and per-CA loop and need to be cleared every CA loop.

Also regarding "Before this change, it was very explicit that CapacityBufferPodListProcessor depends on fake pods registry", this is true, but it is missing the part that that processor was clearing the registry every loop. This behaves badly if there are 2 processors feeding the same registry (which is the case here). If we moved the clearing logic outside of the processors but not in AutoscalingContext, this can be missed during the implementation of those processors. Having it in AutoscalingContext guarantees it is always cleared once and only once per CA-loop.

File: cluster-autoscaler/capacitybuffer/fakepods/registry.go (hunk 2 above)
Comment (walidghallab) [reply cluster-autoscaler/capacitybuffer/fakepods/registry.go:1]: Done

File: cluster-autoscaler/capacitybuffer/registry/fake_pods_registry.go (hunk 4 above)
Comment (walidghallab) [reply cluster-autoscaler/capacitybuffer/registry/fake_pods_registry.go]: Done

File: cluster-autoscaler/processors/capacitybuffer/pod_list_processor.go (hunk 5 above)
Comment (walidghallab) [reply cluster-autoscaler/processors/capacitybuffer/pod_list_processor.go:115]: We need the capacitybuffer itself not just its objectreference.

And keeping it here makes checking capacity buffer for a given capacity buffer pod always exist. If we get rid of it, we:
- Might get 404 error since the capacity buffer can be deleted by the time we go to scale-up status processor.
- If we make API call, it will cost us time. If we don't, we fall into the previous point.

File: cluster-autoscaler/capacitybuffer/fakepods/registry.go (hunk 3 above)
Comment (walidghallab) [reply cluster-autoscaler/capacitybuffer/fakepods/registry.go]: Donw

File: cluster-autoscaler/processors/capacitybuffer/pod_list_processor.go (hunk 5 above)
Comment (norbertcyran) [reply cluster-autoscaler/processors/capacitybuffer/pod_list_processor.go:115]: As far as I see, the buffer object in the scale up status processor is needed only to emit an event. To emit an event, we **do not** need the entire `*v1.CapacityBuffer` object. In fact, the pod and its owner reference have all data that's needed. Event recorder accepts `runtime.Object` and builds `ObjectReference` from it. Since `ObjectReference` itself also implements `runtime.Object`, we can just convert the `OwnerReference` to an `ObjectReference`:

```go
func getBufferReference (pod *corev1.Pod) *corev1.ObjectReference {
  for _, ref := range pod.OwnerReferences {
    if ref.Kind == "CapacityBuffer" {
      return &corev1.ObjectReference{
        Kind: ref.Kind,
        Name: ref.Name,
        UID: ref.UID,
        APIVersion: ref.APIVersion,
        Namespace: pod.Namespace,
      }
    }
  }
  return nil
}
```

And that's it, no API calls, no registry, no worries if the buffer was removed in the meantime

File: cluster-autoscaler/context/autoscaling_context.go (hunk 1 above)
Comment (norbertcyran) [reply cluster-autoscaler/context/autoscaling_context.go]: >it is missing the part that that processor was clearing the registry every loop. This behaves badly if there are 2 processors feeding the same registry (which is the case here). If we moved the clearing logic outside of the processors but not in AutoscalingContext, this can be missed during the implementation of those processors. Having it in AutoscalingContext guarantees it is always cleared once and only once per CA-loop.

Fair point, thanks for clarifying. I'd still lean towards not having the registry at all (see https://github.com/kubernetes/autoscaler/pull/9353#discussion_r2932328260), but that seems fine too

File: cluster-autoscaler/processors/capacitybuffer/pod_list_processor.go (hunk 5 above)
Comment (walidghallab) [reply cluster-autoscaler/processors/capacitybuffer/pod_list_processor.go:115]: Makes sense, didn't know that scale-up status processor only cares about buffer UID.
Thanks Norbert!

File: cluster-autoscaler/context/autoscaling_context.go (hunk 1 above)
Comment (towca) [reply cluster-autoscaler/context/autoscaling_context.go]: I don't have bandwidth to review this PR, but in principle I agree that extending `AutoscalingContext` should not be treated lightly, for the reasons listed by Norbert and more.

If the problem is just having `Registry.Clear()` called exactly once in a known place in `RunOnce()`, this can be solved in other ways that don't require extending the context:
* You can have StaticAutoscaler depend on the object directly (add it to AutoscalerOptions, pass to StaticAutoscaler constructor, save in a field on StaticAutoscaler).
* You can integrate the Registry with LoopStartNotifier, it'll get called near the beginning of RunOnce, before the PodListProcessor.
* You can integrate the Registry with AutoscalingStatusProcessor, it'll get called at the very end of RunOnce.

Comment (walidghallab) [reply cluster-autoscaler/context/autoscaling_context.go]: Thanks Kuba for the suggestions :)

Out of curiosity, should we always avoid AutoscalingContext, or what is good use case for using it?

Comment (walidghallab) [reply cluster-autoscaler/context/autoscaling_context.go]: Adding to last comment, it is worth noting it is being used in at least 4 CA loop processors.
+ benefit of resetting it every CA loop easily.

That's why I thought it was a pretty good use case to be in AutoscalingContext.

File: cluster-autoscaler/processors/capacitybuffer/pod_list_processor.go (hunk 5 above)
Comment (walidghallab) [reply cluster-autoscaler/processors/capacitybuffer/pod_list_processor.go:115]: We actually need the buffer registry in other places.

File: cluster-autoscaler/capacitybuffer/fakepods/registry.go (hunk 3 above)
Comment (walidghallab) [reply cluster-autoscaler/capacitybuffer/fakepods/registry.go]: Rethinking about it, I think my initial decision was good. Bec:
- Registry shouldn't be read asyncronously, bec it gets cleared at the beginning of the CA loop and you might get incorrect result.
- If it is read sync, mutex-free provide slightly more efficiency.
